### PR TITLE
[netcore] Implement AssemblyLoadContext.InternalLoad

### LINKED ISSUE #14830: [netcore]Implement AssemblyLoadContext.InternalLoad
Should probably be grouped w/ collectible assembly work

### PR DESCRIPTION
**Pull request by lambdageek**
Fixes #14830

### CODE REVIEW

File: mono/metadata/appdomain.c [hunk 1]
```diff
@@ -2544,21 +2562,44 @@ ves_icall_System_AppDomain_LoadAssemblyRaw (MonoAppDomainHandle ad,
...
+		MONO_HANDLE_SET (refass, evidence, evidence);
+
+leave:
+	return refass;
+}
+#endif /* ENABLE_NETCORE */
+
+static MonoAssembly*
+mono_alc_load_raw_bytes (MonoAssemblyLoadContext *alc, char *assembly_data, guint32 raw_assembly_len, mono_byte *raw_symbol_data, guint32 raw_symbol_len, gboolean refonly, MonoError *error)
+{
```
Comment (vargaz) [inline mono/metadata/appdomain.c]: Should use guint8* for assembly_data.

File: netcore/System.Private.CoreLib/src/System.Runtime.Loader/AssemblyLoadContext.cs [hunk 2]
```diff
@@ -43,7 +43,16 @@ Assembly InternalLoadFromPath (string assemblyPath, string nativeImagePath)
 
 		internal Assembly InternalLoad (byte[] arrAssembly, byte[] arrSymbols)
 		{
-			throw new NotImplementedException ();
+			Assembly result = null;
```
Comment (marek-safar) [inline netcore/System.Private.CoreLib/src/System.Runtime.Loader/AssemblyLoadContext.cs]: The initialization is redundant

File: netcore/System.Private.CoreLib/src/System.Runtime.Loader/AssemblyLoadContext.cs [hunk 3]
```diff
@@ -43,7 +43,16 @@ Assembly InternalLoadFromPath (string assemblyPath, string nativeImagePath)
...
+			unsafe {
+				int symbolsLength = arrSymbols != null ? arrSymbols.Length : 0;
+				fixed (byte* ptrAssembly = arrAssembly, ptrSymbols = arrSymbols)
+				{
+					result = InternalLoadFromStream (NativeALC, new IntPtr (ptrAssembly), arrAssembly.Length,
```
Comment (marek-safar) [inline netcore/System.Private.CoreLib/src/System.Runtime.Loader/AssemblyLoadContext.cs]: nit: could be just "return ..."

File: netcore/System.Private.CoreLib/src/System.Runtime.Loader/AssemblyLoadContext.cs [hunk 4]
```diff
@@ -43,7 +43,16 @@ Assembly InternalLoadFromPath (string assemblyPath, string nativeImagePath)
 
 		internal Assembly InternalLoad (byte[] arrAssembly, byte[] arrSymbols)
 		{
-			throw new NotImplementedException ();
+			Assembly result = null;
+			unsafe {
+				int symbolsLength = arrSymbols != null ? arrSymbols.Length : 0;
```
Comment (marek-safar) [inline netcore/System.Private.CoreLib/src/System.Runtime.Loader/AssemblyLoadContext.cs]: nit: `int symbolsLength = arrSymbols?.Length ?? 0;`

File: mono/metadata/appdomain.c [hunk 5]
```diff
@@ -2556,21 +2574,44 @@ ves_icall_System_AppDomain_LoadAssemblyRaw (MonoAppDomainHandle ad,
...
+		raw_symbol_data = (mono_byte*) MONO_ARRAY_HANDLE_PIN (raw_symbol_store, mono_byte, 0, &symbol_gchandle);
 	}
 
-	if (!MONO_HANDLE_IS_NULL(raw_symbol_store)) {
-		guint32 symbol_len = mono_array_handle_length (raw_symbol_store);
-		uint32_t symbol_gchandle;
-		mono_byte *raw_symbol_data = (mono_byte*) MONO_ARRAY_HANDLE_PIN (raw_symbol_store, mono_byte, 0, &symbol_gchandle);
-		mono_debug_open_image_from_memory (image, raw_symbol_data, symbol_len);
+	ass = mono_alc_load_raw_bytes (alc, assembly_data, raw_assembly_len, raw_symbol_data, symbol_len, refonly, error);
+	if (symbol_gchandle)
```
Comment (CoffeeFlux) [inline mono/metadata/appdomain.c]: `mono_gchandle_free_internal` already has a null check internally.

File: mono/metadata/appdomain.c [hunk 6]
```diff
@@ -2556,21 +2574,44 @@ ves_icall_System_AppDomain_LoadAssemblyRaw (MonoAppDomainHandle ad,
...
 	MonoAssemblyLoadContext *alc = mono_domain_default_alc (domain);
-	MonoImage *image = mono_image_open_from_data_internal (alc, assembly_data, raw_assembly_len, FALSE, NULL, refonly, FALSE, NULL);
 
-	if (!image) {
-		mono_error_set_bad_image_by_name (error, "In memory assembly", "0x%p", raw_data);
-		return refass;
+	mono_byte *raw_symbol_data = NULL;
+	guint32 symbol_len = 0;
+	uint32_t symbol_gchandle = 0;
+	if (!MONO_HANDLE_IS_NULL(raw_symbol_store)) {
```
Comment (CoffeeFlux) [inline mono/metadata/appdomain.c]: Nit: space

File: mono/metadata/appdomain.c [hunk 7]
```diff
@@ -2556,21 +2574,44 @@ ves_icall_System_AppDomain_LoadAssemblyRaw (MonoAppDomainHandle ad,
...
-		uint32_t symbol_gchandle;
-		mono_byte *raw_symbol_data = (mono_byte*) MONO_ARRAY_HANDLE_PIN (raw_symbol_store, mono_byte, 0, &symbol_gchandle);
-		mono_debug_open_image_from_memory (image, raw_symbol_data, symbol_len);
+	ass = mono_alc_load_raw_bytes (alc, assembly_data, raw_assembly_len, raw_symbol_data, symbol_len, refonly, error);
+	if (symbol_gchandle)
 		mono_gchandle_free_internal (symbol_gchandle);
+	goto_if_nok (error, leave);
+
+	refass = mono_assembly_get_object_handle (domain, ass, error);
+	if (!MONO_HANDLE_IS_NULL(refass))
```
Comment (CoffeeFlux) [inline mono/metadata/appdomain.c]: Nit: space

**Comment by lambdageek** [comment]
@monojenkins build failed

**Comment by lewing** [comment]
@monojenkins build failed

File: netcore/System.Private.CoreLib/src/System.Runtime.Loader/AssemblyLoadContext.cs [hunk 8]
```diff
@@ -43,7 +43,14 @@ Assembly InternalLoadFromPath (string assemblyPath, string nativeImagePath)
 
 		internal Assembly InternalLoad (byte[] arrAssembly, byte[] arrSymbols)
 		{
-			throw new NotImplementedException ();
+			unsafe {
+				int symbolsLength = arrSymbols?.Length ?? 0;
+				fixed (byte* ptrAssembly = arrAssembly, ptrSymbols = arrSymbols)
```
Comment (EgorBo) [inline netcore/System.Private.CoreLib/src/System.Runtime.Loader/AssemblyLoadContext.cs]: wow, TIL: you can fix multiple arrays in one `fixed` block 🙂

**Comment by lambdageek** [comment]
@monojenkins build

**Pull request merged**